06_gpu_and_ml/audio-to-text/parakeet.py

```diff
@@ -0,0 +1,246 @@
+# We'll show you how to use Parakeet for real-time audio transcription,
+# with a simple python client and a websocket server you can spin up easily in Modal.
+
+# This example uses the `nvidia/parakeet-tdt-0.6b-v2` model, which, as of May 13, 2025, sits at the
+# top of Hugging Face's [ASR leaderboard](https://huggingface.co/spaces/hf-audio/open_asr_leaderboard).
+
+# To run this example:
+
+# 1. Start the server locally:
+# ```bash
+# modal serve 06_gpu_and_ml/audio-to-text/parakeet.py::server_app
+# ```
+
+# 2. In a separate terminal, run the client:
+# ```bash
+# modal run 06_gpu_and_ml/audio-to-text/parakeet.py::client_app --modal-profile=$(modal profile current)
+# ```
+
+# See [Troubleshooting](https://modal.com/docs/examples/parakeet#client) at the bottom if you run into issues.
+
+# Here's what your final output might look like:
+
+# ```
+# ☀️ Waking up model, this may take a few seconds on cold start...
+
+# Recording and streaming... Press Ctrl+C to stop.
+# 📝 Transcription: Hi, how's it going?.
+# 📝 Transcription: Doing well. Great day to be having fun with transcription models
+# 📝 Transcription:
+# 📝 Transcription: Good.
+# ^C
+# 🛑 Stopped by user.
+# ```
+
+# ## Setup
+import modal
+
+app_name = "parakeet-websocket"
+
+server_app = modal.App(app_name)
+BUFFER_SIZE = 8000
+
+# ## Volume for caching model weights
+# We use a [Modal Volume](https://modal.com/docs/guide/volumes) to cache the model weights.
+# This allows us to avoid downloading the model weights every time we start a new instance.
+
+
+model_cache = modal.Volume.from_name("parakeet-model-cache", create_if_missing=True)
+# ## Configuring dependencies
+# The model runs remotely inside a [custom container](https://modal.com/docs/guide/custom-container). We can define the environment
+# and install our Python dependencies in that container's `Image`.
+
+# For inference, we recommend using the official NVIDIA CUDA Docker images from Docker Hub.
+# You'll need to add Python 3 and pip with the `add_python` option because the image
+# doesn't have these by default.
+
+# Additionally, we install `ffmpeg` for handling audio data and `fastapi` to create a web
+# server for our websocket. `uv` gives us even more speed on installation!
+
+image = (
+    modal.Image.from_registry(
+        "nvidia/cuda:12.8.0-cudnn-devel-ubuntu22.04", add_python="3.12"
+    )
+    .pip_install("uv")
+    .env(
+        {
+            "HF_HUB_ENABLE_HF_TRANSFER": "1",
+            "HF_HOME": "/cache",  # cache directory for Hugging Face models
+            "DEBIAN_FRONTEND": "noninteractive",
+            "CXX": "g++",
+            "CC": "g++",
+        }
+    )
+    .apt_install("ffmpeg")
+    .run_commands(
+        "uv pip install --system hf_transfer==0.1.9 huggingface_hub[hf-xet]==0.31.2 nemo_toolkit[asr]==2.3.0 cuda-python==12.9.0",
```

Comment:
Curious -- why are we using uv pip install here? And in multiple steps?

Reply:
> why are we using uv pip install here?

Because it's faster than pip. Maybe I'm not understanding the question. 

> And in multiple steps?

The additional numpy step is there to downgrade numpy. The current numpy version installed with NeMo is incompatible with this model.

Reply:
> why are we using uv pip install here?

I don't think we usually use `uv` in examples, even though it is faster, because we don't have an interface for it, so it feels a bit janky: `run_commands([...])` instead of `pip_install("", ...)`.

> The additional numpy step is there to downgrade numpy. The current numpy version installed with NeMo is incompatible with this model.

Why can't we do `numpy<2` within the same step? And I don't think we need a third step to install `fastapi`.

Asking because `pip` doesn't take into account dependency resolution across invocations, and I don't think `uv pip` does either.

Reply:
Also I think some of the speedup may be illusory because deferred to a runtime expense -- by default, `uv` does not compile bytecode for packages. cf [Image guide](https://modal.com/docs/guide/images#run-shell-commands-with-run_commands) in the docs

![Screenshot_20250515-180014.png](https://github.com/user-attachments/assets/bd360ef7-303e-40bb-93c0-50ea1e665b58)

Reply:
removed uv